platform/default/src/mbgl/storage/pmtiles_file_source.cpp

```diff
@@ -0,0 +1,547 @@
+                                tileResponse.error->reason,
+                                std::string("Error fetching PMTiles tile: ") + tileResponse.error->message);
+                            ref.invoke(&FileSourceRequest::setResponse, response);
+                            return;
+                        }
+
+                        response.data = tileResponse.data;
+                        response.noContent = false;
+                        response.modified = tileResponse.modified;
+                        response.expires = tileResponse.expires;
+                        response.etag = tileResponse.etag;
+
+                        if (header.tile_compression == pmtiles::COMPRESSION_GZIP) {
+                            response.data = std::make_shared<std::string>(util::decompress(*tileResponse.data));
+                        }
+
+                        ref.invoke(&FileSourceRequest::setResponse, response);
+                        return;
+                    });
+                });
+        });
+    }
+
+    void setResourceOptions(ResourceOptions options) {
+        std::lock_guard<std::mutex> lock(resourceOptionsMutex);
+        resourceOptions = options;
+    }
+
+    ResourceOptions getResourceOptions() {
+        std::lock_guard<std::mutex> lock(resourceOptionsMutex);
+        return resourceOptions.clone();
+    }
+
+    void setClientOptions(ClientOptions options) {
+        std::lock_guard<std::mutex> lock(clientOptionsMutex);
+        clientOptions = options;
+    }
+
+    ClientOptions getClientOptions() {
+        std::lock_guard<std::mutex> lock(clientOptionsMutex);
+        return clientOptions.clone();
+    }
+
+private:
+    mutable std::mutex resourceOptionsMutex;
+    mutable std::mutex clientOptionsMutex;
+    ResourceOptions resourceOptions;
+    ClientOptions clientOptions;
+
+    std::shared_ptr<FileSource> fileSource;
+    std::map<std::string, pmtiles::headerv3> header_cache;
+    std::map<std::string, std::string> metadata_cache;
+    std::map<std::string, std::map<std::string, std::vector<pmtiles::entryv3>>> directory_cache;
+    std::map<std::string, std::vector<std::string>> directory_cache_control;
+    std::map<AsyncRequest*, std::unique_ptr<AsyncRequest>> tasks;
+
+    std::shared_ptr<FileSource> getFileSource() {
+        if (!fileSource) {
+            fileSource = FileSourceManager::get()->getFileSource(
+                FileSourceType::ResourceLoader, resourceOptions, clientOptions);
+        }
+
+        return fileSource;
+    }
+
+    void getHeaderAndRootDirectory(const std::string& url, AsyncRequest* req, AsyncCallback callback) {
+        if (header_cache.find(url) != header_cache.end()) {
+            callback(std::unique_ptr<Response::Error>());
+        }
+
+        Resource resource(Resource::Kind::Source, url);
+        resource.loadingMethod = Resource::LoadingMethod::Network;
+        resource.dataRange = std::make_pair<uint64_t, uint64_t>(0, 16383);
```

Comment:
Avoid magic numbers

Reply:
The numbers are following the PMTiles specification (same as the [JS version](https://github.com/protomaps/PMTiles/blob/main/js/index.ts#L613)). Should I just put a comment telling about their origin, or should I make something different in this piece of code? I'm open to advice.

Reply:
I make some changes to the code, and a comment explaining where does the numbers come from, could you please review?

Reply:
Adding comments is good, but I would also assign these numbers to a const.

Reply:
That's a very good idea. Please check if it's better now.